BinaryFormatter Migration Guide (#41564)
dotnet/docs: docs/standard/serialization/binaryformatter-migration-guide/winforms-applications.md

Reviewer: @lonitra / @JeremyKuhne - I applied my own feedback directly to this page, but please feel free to refine further as needed. I applied the same type of edits to the wpf-applications page too.

/cc @terrajobst @gewarren

---
title: "BinaryFormatter migration guide: WinForms applications"
description: "Learn about the effects of the deprecation and removal of BinaryFormatter from .NET on Windows Forms and how to migrate."
ms.date: 7/31/2024
no-loc: [BinaryFormatter, WinForms]
helpviewer_keywords:
  - "BinaryFormatter"
  - "WinForms"
---

# Windows Forms migration guide for BinaryFormatter

## BinaryFormatter removal

> Starting in .NET 9, `BinaryFormatter` has moved to a separate, unsupported NuGet package due to its known [security risks](../binaryformatter-security-guide.md). For more information about the risks `BinaryFormatter` poses and the reason for its removal, see [BinaryFormatter migration guide](overview.md). With BinaryFormatter’s removal, it's expected that many Windows Forms applications will be impacted, and you'll need to take action to complete your migration to .NET 9 or a later version.

Author: ```suggestion
Starting in .NET 9, `BinaryFormatter` has moved to a separate, unsupported NuGet package due to its known [security risks](../binaryformatter-security-guide.md). For more information about the risks `BinaryFormatter` poses and the reason for its removal, see [BinaryFormatter migration guide](index.md). With BinaryFormatter’s removal, it's expected that many Windows Forms applications will be impacted, and you'll need to take action to complete your migration to .NET 9 or a later version.
```

Windows Forms and WPF apps both handle the following types, along with arrays and lists of these types. Clipboard, drag-and-drop, and design-time resources will continue to work with these types without any migration steps needed.

- `bool`
- `byte`
- `char`
- `decimal`
- `double`
- `int`
- `sbyte`
- `float`
- `TimeSpan`
- `DateTime`
- `uint`
- `string`
- `nint`
- `nuint`
- `long`
- `ulong`
- `short`
- `ushort`
- `PointF`
- `RectangleF`

Windows Forms also supports the following additional types:

- `Bitmap`
- `ImageListStreamer`

### OLE scenarios

For information about the effects BinaryFormatter removal has on OLE scenarios such as clipboard and drag-and-drop as well as migration guidance see [Windows Forms and Windows Presentation Foundation BinaryFormatter OLE guidance](./winforms-wpf-ole-guidance.md).

### Resources (ResX)

#### The Windows Forms Designer

The Windows Forms Out-Of-Process Designer also uses `BinaryFormatter` internally for ResX serialization and deserialization.

> Types and properties might participate in serialization without you realizing due to the standard behavior of the Windows Forms Designer. One way that BinaryFormatter is used that you might not be aware of is when a `public` property on a <xref:System.ComponentModel.IComponent> is introduced and that component is populated or that property is edited at design time. It's likely that that data gets serialized into resource files. Consider the following conditions:

Reviewer 2: ```suggestion
Types and properties might participate in serialization without you realizing due to the standard behavior of the Windows Forms Designer. One way that BinaryFormatter is used that you might not be aware of is when a `public` property on a <xref:System.ComponentModel.IComponent> is introduced and that property is populated or edited at design time. That property is serialized into resource files under the following conditions:
```

- A public property contains data at the time when a `Form` in the Designer is saved.
- That property is not read-only.
- That property is not attributed with `DesignerSerializationVisibility(false)`.
- That property does not have a DefaultValueAttribute.
- That property does not have a respective `bool ShouldSerialize[PropertyName]` method that returns `false` at the time of the CodeDOM serialization process. (Note: the method can have `private` scope.)

> - That property does not have a [`DesignerSerializer`](/dotnet/api/microsoft.visualstudio.modeling.dsldefinition.designerserializer)

Reviewer 3: @Tanya-Solyanik this is where new bullet was added based off of your previous suggestion

> - That property does not have a [`DesignerSerializer`](/dotnet/api/microsoft.visualstudio.modeling.dsldefinition.designerserializer)

Reviewer 2: ```suggestion
- That property is of type that does not have a [`DesignerSerializer`](/dotnet/api/microsoft.visualstudio.modeling.dsldefinition.designerserializer)
```

That attribute usually decorates the type.

- A public property contains data at the time when a `Form` in the Designer is saved.
- That property is not read-only.
- That property is not attributed with `DesignerSerializationVisibility(false)`.
- That property does not have a DefaultValueAttribute.
- That property does not have a respective `bool ShouldSerialize[PropertyName]` method that returns `false` at the time of the CodeDOM serialization process. (Note: the method can have `private` scope.)
- That property does not have a [`DesignerSerializer`](/dotnet/api/microsoft.visualstudio.modeling.dsldefinition.designerserializer)

If these statements are true, the Designer determines if that property’s type has a type converter. If it does, the Designer uses the type converter to serialize the property content. Otherwise, it uses BinaryFormatter to serialize the content into the resource file.
Windows Forms has added analyzers along with code fixes to help bring awareness to this type of behavior where BinaryFormatter serialization might be occurring without the developer’s knowledge.

#### Loading resource during runtime

> Types that had been previously serialized into resource files via `BinaryFormatter` will continue to deserialize as expected without the need for `BinaryFormatter` as the content of ResX files are considered trusted data. In the rare case that deserialization cannot occur without `BinaryFormatter`, it can be added back for compatibility. See [BinaryFormatter migration guide: Compatibility Package](compatibility-package.md) for details. Note that an extra step of setting `System.Resources.Extensions.UseBinaryFormatter` app context switch to `true` is required to use `BinaryFormatter` for resources.

Reviewer 3: @jeffhandley I believe this is about the extent that users need to know regarding embedded resources and WinForms, so I'm not sure a dedicated page is needed for this. Did we want to provide in depth details around embedded resources?

Reviewer 4: Can you clarify what you mean by this? Are you suggesting that the entire page should be collapsed to this one section and that this section could live elsewhere?

Reviewer 3: In the outline provided in PR description we have dedicated page for embedded resources that is assigned to Jeremy and I. In the context of WinForms, this is all the user needs to know about their embedded resources -- that we have taken care of it so that resources will still deserialize as expected without BinaryFormatter and should users run into a case where it is not deserializing as expected, the compat package needs to be pulled in. But I am not sure if what was intended for that page to begin with was to be broader than just WinForms and have more in-depth technical information.

Reviewer 4: Ah, I see. That makes sense.

I've added a section to index.md that points to this section.
